app/javascript/src/NewService/components/FormTemplate.jsx

```diff
@@ -0,0 +1,145 @@
+// @flow
+
+import React from 'react'
+import {CSRFToken} from 'utilities/utils'
+import type {FormTemplateProps} from 'NewService/types'
+
+type ServiceDiscoveryFormTemplateProps = {
+  hiddenServiceDiscoveryInput: {
+    value: string,
+    name: string,
+    id: string
+  },
+  serviceDiscoverySelect: {
+    [string]: {
+        htmlFor: string,
+        label: string,
+        name: string,
+        id: string
+    }
+  }
+}
+
+const serviceDiscoveryProps: ServiceDiscoveryFormTemplateProps = {
+  hiddenServiceDiscoveryInput: {
+    value: 'discover',
+    name: 'service[source]',
+    id: 'service_source'
+  },
+  serviceDiscoverySelect: {
+    project: {
+      htmlFor: 'namespace',
+      label: 'Namespace',
+      name: 'service[namespace]',
+      id: 'service_namespace'
+    },
+    service: {
+      htmlFor: 'service_name',
+      label: 'Name',
+      name: 'service[name]',
+      id: 'service_name'
+    }
+  }
+}
+
+const Label = (
+  project: {htmlFor: string, label: string}, required: boolean = false
+) =>
+  <label htmlFor={project.htmlFor}>{project.label}
+    { required && <abbr title="required">*</abbr> }
+  </label>
+
+const SelectOptions = (options: string[]) =>
+  options.map(option => <option key={option} value={option}>{option}</option>)
+
+const Select = (
+  project: {name: string, id: string}, props: FormTemplateProps
+) =>
+  <select required="required" name={project.name}
+    id={project.id} onChange={props.onChangeNamespaces}
+    data-selected-value={
+      project.id === 'service_namespace' ? props.selectedNamespace : ''
+    }>
+    {props.namespaces && SelectOptions(props.namespaces)}
+  </select>
+
+const ErrorMessage = ({error}) =>
+  <p className="errorMessage">
+    { error.errorMsg &&
+      `Sorry, your request has failed with the error: ${error.errorMsg}`
+    }
+  </p>
+
+const serviceDiscoveriItems = (props, serviceDiscoveryProps) => {
+  return (
+    <React.Fragment>
+      { props.error && <ErrorMessage error={props.error}/>}
+      <li id="service_name_input" className="string required">
+        { Label(serviceDiscoveryProps.serviceDiscoverySelect.project, false) }
```

Comment:
Not convinced. This approach leads you to create a lot of small components that maybe are used only in a single file (like here). We should avoid creating unnecessary components, I don't like to over-engineering just because is the `React way` (whatever it means). If needed, it can be extracted to a new component in the future.

Reply:
In react, components are functions and it works the other way around. Therefore, you don't have to create anything new, you have it already: `Label`. It's the way one render or call those functions what matters and in this particular case `Label` as a vanilla function makes it harder to read.

Reply:
Yeah, this is true. It's going to be better and uniform. Also, it might easily solve what I commented here: https://github.com/3scale/porta/pull/763/files#r285061091